rfcs/text/0000-multiple-users.md

```diff
@@ -0,0 +1,419 @@
+-->
+
+<!--
+Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting.
+-->
+
+<!--
+Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
+-->
+
+## Scope of impact
+
+### New fields for IAM
+
+The fields `user.[changes|effective|target].*` are net new fields,
+so they don't represent a breaking change. They are especially important
+for security-related data sources around IAM and audit logs.
+These event sources should be adjusted to populate these new fields, as they are
+very important in getting a complete picture of user management activity.
+
+Some event sources for user management activity may have used `user.*` fields at the
+root to describe the user being modified, rather than the user performing the action.
+These sources will have to be modified to be consistent with the fact that user
+fields at the root are meant to represent who's performing the action.
+
+### New user field duplication guidance
+
+In order to firmly establish the user fields at the root of the event as the user
+performing the action, this RFC introduces new guidance:
+
+* Remote logons `source.user.*` should be copied to `user.*`
+* Purpose-specific fields such as `url.username` should be copied at `user.name`
+
+These came up while working on this RFC; this is not guidance that was given
+in the past. Data sources that populate these fields will need to be revisited
+and adjusted accordingly.
+
+<!-- TODO
+
+Depending on the outcome of the discussion on `host.user.*`, mention it here.
+It's currently listed in the concerns below.
+
+-->
+
+## Concerns
+
+### Deprecating host.user fields
+
+In past discussions and recent research, we have not identified a clear purpose
+for the user fields nested at `host.user.*`.
+
+We are considering deprecating these fields with the intent to remove them completely.
+Please let us know if you disagree with this, and share how you're using them.
+
+#### Resolution
+
+No resolution yet.
+
+### Documenting the purpose of each usage of the user fields
+
+As of ECS 1.6, the ECS documentation doesn't have a good place to explain at length
+how to properly use the multiple nesting locations for `user`.
+This is already a problem for the usage of `user` at the root vs its 5 reuse locations.
+The addition of 3 new reuse locations adds to this situation.
+
+#### Resolution
+
+Adding a way to document field sets via free form text is being worked on independently
+of this proposal ([ecs#943](https://github.com/elastic/ecs/issues/943)).
```

Comment:
I wonder if it'd also be useful to have some `description` field or maybe `purpose` within the reusable section to document why the reuse is happening:

```YAML
# schemas/user.yml excerpt
  reusable:
    top_level: true
    expected:
      - client
      - destination
      - host
      - server
      - source
      # Added for this RFC
      - at: user
        as: target
        # field for documenting why the reuse is happening
        description: Whenever a user is performing an action that affects another user
          -- typically in IAM scenarios -- the user affected by the action is captured at user.target
      - at: user
        as: effective
      - at: user
        as: changes
```

Reply:
Yes we will handle that a bit more loosely. We have work in progress (https://github.com/elastic/ecs/issues/932) to support a section of free form asciidoc documentation per field set (in addition to all of the attributes in the yaml).

Once we have the mechanism for this free form documentation section, the "user" field set documentation will lay out all expected uses with examples, based on what's in this RFC.

Reply:
Nice! Sounds good.